crates/python/qcs_sdk/_execution_data.pyi

```diff
@@ -6,67 +6,142 @@ It is only here to represent the structure of the rust source code 1:1
 import datetime
 from typing import List, Optional
 
-class _IntegerReadoutValues:
+import numpy as np
+from numpy.typing import NDArray
+
+from crates.python.qcs_sdk.qpu.result_data import QPUResultData
+from crates.python.qcs_sdk.qvm.result_data import QVMResultData
+
+class RegisterMatrixConversionError(ValueError):
+    """Error that may occur when building a ``RegisterMatrix`` from execution data."""
+
+    ...
+
+class RegisterMatrix:
+    """
+    Values in a 2-dimensional ``ndarray`` representing the final shot value in each memory reference across all shots.
+    Each variant corresponds to the possible data types a register can contain.
+
+    Variants:
+        ``integer``: Corresponds to the Quil `BIT`, `OCTET`, or `INTEGER` types.
+        ``real``: Corresponds to the Quil `REAL` type.
+        ``complex``: Registers containing complex numbers.
+
+    Methods (each per variant):
+        - ``is_*``: if the underlying values are that type.
+        - ``as_*``: if the underlying values are that type, then those values, otherwise ``None``.
+        - ``to_*``: the underlying values as that type, raises ``ValueError`` if they are not.
+        - ``from_*``: wrap underlying values as this enum type.
+
+    """
+
+    def is_integer(self) -> bool: ...
+    def is_real(self) -> bool: ...
+    def is_complex(self) -> bool: ...
+    def as_integer(self) -> Optional[NDArray[np.int64]]: ...
+    def as_real(self) -> Optional[NDArray[np.float64]]: ...
+    # In numpy `complex128` is a complex number made up of two `f64`s.
+    def as_complex(self) -> Optional[NDArray[np.complex128]]: ...
+    def to_integer(self) -> Optional[NDArray[np.int64]]: ...
+    def to_real(self) -> Optional[NDArray[np.float64]]: ...
+    def to_complex(self) -> Optional[NDArray[np.complex128]]: ...
```

Comment:
I think the `to_` variants are not `Optional` since they throw if that type isn't found, right?

```suggestion
    def to_integer(self) -> NDArray[np.int64]: ...
    def to_real(self) -> NDArray[np.float64]: ...
    def to_complex(self) -> NDArray[np.complex128]: ...
```